### ISSUE TITLE
Weird CORS problem with superagent and cloudfront amazon S3 in firefox

### ISSUE THREAD

**Issue by nguyenxuantuong** [visionmedia/superagent#574]
I got a very weird error with superagent when making simple request to amazon cloudfront:

There is simple code block:

superagent
  .get('https://danesrti8sy2q.cloudfront.net/public/issues/6491/deploy.json')    
  .set('Accept', 'application/json')
  .end(function(error, res){
  	console.log("res");
  });

In firefox, it will throws the errors:

"NetworkError: 403 Forbidden - https://danesrti8sy2q.cloudfront.net/public/issues/6491/deploy.json"
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://danesrti8sy2q.cloudfront.net/public/issues/6491/deploy.json. This can be fixed by moving the resource to the same domain or enabling CORS.

But using jquery is totally fine:
$.getJSON("https://danesrti8sy2q.cloudfront.net/public/issues/6491/deploy.json", function(data){console.log(data);})

Note that, cloudfront has been configured to accept CORS request. I guess there might be some differences in term of headers btw the one using by jquery and superagent. Any one have an idea?

**Comment by defunctzombie** [comment]
Does it work fine in other browsers?

**Comment by nguyenxuantuong** [comment]
Yes. From what i test, the error only occur in firefox.

**Comment by rdy** [comment]
I believe this is the same issue as #484, there is an onprogress callback which is added to the xhr. Doing this on a cross domain request causes a security exception only in firefox.

I've modified this on my own fork and removed the callback to work around the issue 9b1e68e4036f12d01b7172141f0c6a2c44d546bf, it looks like firefox is not going to change this. Instead it needs detect the cross origin and not bind the callback in that case.

**Comment by Mech7** [comment]
Am having the same problem :(

**Comment by dlockhart** [comment]
I also ran into this issue. Following the trail @rdy suggested, I found this ominous defect on Bugzilla: [Unnecessary CORS preflight when using upload progress](https://bugzilla.mozilla.org/show_bug.cgi?id=727412).

Mozilla claims it's not a defect at all, specifically referencing the [XmlHttpRequest spec](https://xhr.spec.whatwg.org/) which states that it will it will use "CORS-with-forced-preflight if one or more event listeners are registered on the associated XMLHttpRequestUpload object...".

So it appears as though when `superagent` registers for the onprogress event:
```javascript
// client.js
if (xhr.upload) {
  xhr.upload.onprogress = function(e){
    e.percent = e.loaded / e.total * 100;
    self.emit('progress', e);
  };
}
```
This is causing Firefox to do a preflight `OPTIONS` request, which in our case is getting blocked by cloudfront (not your problem). Chrome, IE and Safari seem to have a different interpretation of the spec, and aren't doing the prelight.

@defunctzombie, is there maybe a way to not register for the `onprogress` event if it's not needed (e.g. no event listeners have been added)? Because even if we configure cloudfront to allow the `OPTIONS` requests through, there will still be one for every request we make through `superagent` (on Firefox), even just simple `GET` requests that shouldn't require a preflight.
